File: jump/with_symmetry_elimination_constraints/thomson_8_z_ordering.jl
# NLP written by GAMS Convert at 03/17/24 14:47:55
#
# Equation counts
#     Total        E        G        L        N        X        C        B
#        15        8        7        0        0        0        0        0
#
# Variable counts
#                  x        b        i      s1s      s2s       sc       si
#     Total     cont   binary  integer     sos1     sos2    scont     sint
#        24       24        0        0        0        0        0        0
# FX      4
#
# Nonzero counts
#     Total    const       NL
#        38       14       24
#
# Reformulation has removed 1 variable and 1 equation

using JuMP
using MathOptInterface
const MOI = MathOptInterface

model = m = Model()

@variable(m, 0 <= x1 <= 0, start=0)
@variable(m, 0 <= x2 <= 1, start=0)
@variable(m, -1 <= x3 <= 1, start=0)
@variable(m, -1 <= x4 <= 1, start=0)
@variable(m, -1 <= x5 <= 1, start=0)
@variable(m, -1 <= x6 <= 1, start=0)
@variable(m, -1 <= x7 <= 1, start=0)
@variable(m, -1 <= x8 <= 1, start=0)
@variable(m, 0 <= x9 <= 0, start=0)
@variable(m, 0 <= x10 <= 0, start=0)
@variable(m, -1 <= x11 <= 1, start=0)
@variable(m, -1 <= x12 <= 1, start=0)
@variable(m, -1 <= x13 <= 1, start=0)
@variable(m, -1 <= x14 <= 1, start=0)
@variable(m, -1 <= x15 <= 1, start=0)
@variable(m, -1 <= x16 <= 1, start=0)
@variable(m, 1 <= x17 <= 1, start=1)
@variable(m, -1 <= x18 <= 1, start=0)
@variable(m, -1 <= x19 <= 1, start=0)
@variable(m, -1 <= x20 <= 1, start=0)
@variable(m, -1 <= x21 <= 1, start=0)
@variable(m, -1 <= x22 <= 1, start=0)
@variable(m, -1 <= x23 <= 1, start=0)
@variable(m, -1 <= x24 <= 1, start=0)

@NLobjective(m, Min, 1 / sqrt((x1 - x2)^2 + (x9 - x10)^2 + (x17 - x18)^2) + 1
    / sqrt((x1 - x3)^2 + (x9 - x11)^2 + (x17 - x19)^2) + 1 / sqrt((x1 - x4)^2
    + (x9 - x12)^2 + (x17 - x20)^2) + 1 / sqrt((x1 - x5)^2 + (x9 - x13)^2 + (
    x17 - x21)^2) + 1 / sqrt((x1 - x6)^2 + (x9 - x14)^2 + (x17 - x22)^2) + 1 /
    sqrt((x1 - x7)^2 + (x9 - x15)^2 + (x17 - x23)^2) + 1 / sqrt((x1 - x8)^2 + (
    x9 - x16)^2 + (x17 - x24)^2) + 1 / sqrt((x2 - x3)^2 + (x10 - x11)^2 + (x18
    - x19)^2) + 1 / sqrt((x2 - x4)^2 + (x10 - x12)^2 + (x18 - x20)^2) + 1 /
    sqrt((x2 - x5)^2 + (x10 - x13)^2 + (x18 - x21)^2) + 1 / sqrt((x2 - x6)^2 +
    (x10 - x14)^2 + (x18 - x22)^2) + 1 / sqrt((x2 - x7)^2 + (x10 - x15)^2 + (
    x18 - x23)^2) + 1 / sqrt((x2 - x8)^2 + (x10 - x16)^2 + (x18 - x24)^2) + 1
    / sqrt((x3 - x4)^2 + (x11 - x12)^2 + (x19 - x20)^2) + 1 / sqrt((x3 - x5)^2
    + (x11 - x13)^2 + (x19 - x21)^2) + 1 / sqrt((x3 - x6)^2 + (x11 - x14)^2 +
    (x19 - x22)^2) + 1 / sqrt((x3 - x7)^2 + (x11 - x15)^2 + (x19 - x23)^2) + 1
    / sqrt((x3 - x8)^2 + (x11 - x16)^2 + (x19 - x24)^2) + 1 / sqrt((x4 - x5)^2
    + (x12 - x13)^2 + (x20 - x21)^2) + 1 / sqrt((x4 - x6)^2 + (x12 - x14)^2 +
    (x20 - x22)^2) + 1 / sqrt((x4 - x7)^2 + (x12 - x15)^2 + (x20 - x23)^2) + 1
    / sqrt((x4 - x8)^2 + (x12 - x16)^2 + (x20 - x24)^2) + 1 / sqrt((x5 - x6)^2
    + (x13 - x14)^2 + (x21 - x22)^2) + 1 / sqrt((x5 - x7)^2 + (x13 - x15)^2 +
    (x21 - x23)^2) + 1 / sqrt((x5 - x8)^2 + (x13 - x16)^2 + (x21 - x24)^2) + 1
    / sqrt((x6 - x7)^2 + (x14 - x15)^2 + (x22 - x23)^2) + 1 / sqrt((x6 - x8)^2
    + (x14 - x16)^2 + (x22 - x24)^2) + 1 / sqrt((x7 - x8)^2 + (x15 - x16)^2 +
    (x23 - x24)^2))

@NLconstraint(m, e1, x1^2 + x9^2 + x17^2 == 1)
@NLconstraint(m, e2, x2^2 + x10^2 + x18^2 == 1)
@NLconstraint(m, e3, x3^2 + x11^2 + x19^2 == 1)
@NLconstraint(m, e4, x4^2 + x12^2 + x20^2 == 1)
@NLconstraint(m, e5, x5^2 + x13^2 + x21^2 == 1)
@NLconstraint(m, e6, x6^2 + x14^2 + x22^2 == 1)
@NLconstraint(m, e7, x7^2 + x15^2 + x23^2 == 1)
@NLconstraint(m, e8, x8^2 + x16^2 + x24^2 == 1)
@constraint(m, e9, x17 - x18 >= 0)
@constraint(m, e10, x18 - x19 >= 0)
@constraint(m, e11, x19 - x20 >= 0)
@constraint(m, e12, x20 - x21 >= 0)
@constraint(m, e13, x21 - x22 >= 0)
@constraint(m, e14, x22 - x23 >= 0)
@constraint(m, e15, x23 - x24 >= 0)

File: pyomo/with_symmetry_elimination_constraints/thomson_8_z_ordering.py
# NLP written by GAMS Convert at 03/17/24 14:47:59
#
# Equation counts
#     Total        E        G        L        N        X        C        B
#        15        8        7        0        0        0        0        0
#
# Variable counts
#                  x        b        i      s1s      s2s       sc       si
#     Total     cont   binary  integer     sos1     sos2    scont     sint
#        24       24        0        0        0        0        0        0
# FX      4
#
# Nonzero counts
#     Total    const       NL
#        38       14       24
#
# Reformulation has removed 1 variable and 1 equation

from pyomo.environ import *

model = m = ConcreteModel()

m.x1 = Var(within=Reals, bounds=(0,0), initialize=0)
m.x2 = Var(within=Reals, bounds=(0,1), initialize=0)
m.x3 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x4 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x5 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x6 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x7 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x8 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x9 = Var(within=Reals, bounds=(0,0), initialize=0)
m.x10 = Var(within=Reals, bounds=(0,0), initialize=0)
m.x11 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x12 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x13 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x14 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x15 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x16 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x17 = Var(within=Reals, bounds=(1,1), initialize=1)
m.x18 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x19 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x20 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x21 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x22 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x23 = Var(within=Reals, bounds=(-1,1), initialize=0)
m.x24 = Var(within=Reals, bounds=(-1,1), initialize=0)

m.obj = Objective(sense=minimize, expr= 1 / sqrt((m.x1 - m.x2)**2 + (m.x9 -
    m.x10)**2 + (m.x17 - m.x18)**2) + 1 / sqrt((m.x1 - m.x3)**2 + (m.x9 - m.x11)
    **2 + (m.x17 - m.x19)**2) + 1 / sqrt((m.x1 - m.x4)**2 + (m.x9 - m.x12)**2
    + (m.x17 - m.x20)**2) + 1 / sqrt((m.x1 - m.x5)**2 + (m.x9 - m.x13)**2 + (
    m.x17 - m.x21)**2) + 1 / sqrt((m.x1 - m.x6)**2 + (m.x9 - m.x14)**2 + (m.x17
    - m.x22)**2) + 1 / sqrt((m.x1 - m.x7)**2 + (m.x9 - m.x15)**2 + (m.x17 -
    m.x23)**2) + 1 / sqrt((m.x1 - m.x8)**2 + (m.x9 - m.x16)**2 + (m.x17 - m.x24)
    **2) + 1 / sqrt((m.x2 - m.x3)**2 + (m.x10 - m.x11)**2 + (m.x18 - m.x19)**2)
    + 1 / sqrt((m.x2 - m.x4)**2 + (m.x10 - m.x12)**2 + (m.x18 - m.x20)**2) + 1
    / sqrt((m.x2 - m.x5)**2 + (m.x10 - m.x13)**2 + (m.x18 - m.x21)**2) + 1 /
    sqrt((m.x2 - m.x6)**2 + (m.x10 - m.x14)**2 + (m.x18 - m.x22)**2) + 1 / sqrt
    ((m.x2 - m.x7)**2 + (m.x10 - m.x15)**2 + (m.x18 - m.x23)**2) + 1 / sqrt((
    m.x2 - m.x8)**2 + (m.x10 - m.x16)**2 + (m.x18 - m.x24)**2) + 1 / sqrt((m.x3
    - m.x4)**2 + (m.x11 - m.x12)**2 + (m.x19 - m.x20)**2) + 1 / sqrt((m.x3 -
    m.x5)**2 + (m.x11 - m.x13)**2 + (m.x19 - m.x21)**2) + 1 / sqrt((m.x3 - m.x6)
    **2 + (m.x11 - m.x14)**2 + (m.x19 - m.x22)**2) + 1 / sqrt((m.x3 - m.x7)**2
    + (m.x11 - m.x15)**2 + (m.x19 - m.x23)**2) + 1 / sqrt((m.x3 - m.x8)**2 + (
    m.x11 - m.x16)**2 + (m.x19 - m.x24)**2) + 1 / sqrt((m.x4 - m.x5)**2 + (
    m.x12 - m.x13)**2 + (m.x20 - m.x21)**2) + 1 / sqrt((m.x4 - m.x6)**2 + (
    m.x12 - m.x14)**2 + (m.x20 - m.x22)**2) + 1 / sqrt((m.x4 - m.x7)**2 + (
    m.x12 - m.x15)**2 + (m.x20 - m.x23)**2) + 1 / sqrt((m.x4 - m.x8)**2 + (
    m.x12 - m.x16)**2 + (m.x20 - m.x24)**2) + 1 / sqrt((m.x5 - m.x6)**2 + (
    m.x13 - m.x14)**2 + (m.x21 - m.x22)**2) + 1 / sqrt((m.x5 - m.x7)**2 + (
    m.x13 - m.x15)**2 + (m.x21 - m.x23)**2) + 1 / sqrt((m.x5 - m.x8)**2 + (
    m.x13 - m.x16)**2 + (m.x21 - m.x24)**2) + 1 / sqrt((m.x6 - m.x7)**2 + (
    m.x14 - m.x15)**2 + (m.x22 - m.x23)**2) + 1 / sqrt((m.x6 - m.x8)**2 + (
    m.x14 - m.x16)**2 + (m.x22 - m.x24)**2) + 1 / sqrt((m.x7 - m.x8)**2 + (
    m.x15 - m.x16)**2 + (m.x23 - m.x24)**2))

m.e1 = Constraint(expr= m.x1**2 + m.x9**2 + m.x17**2 == 1)
m.e2 = Constraint(expr= m.x2**2 + m.x10**2 + m.x18**2 == 1)
m.e3 = Constraint(expr= m.x3**2 + m.x11**2 + m.x19**2 == 1)
m.e4 = Constraint(expr= m.x4**2 + m.x12**2 + m.x20**2 == 1)
m.e5 = Constraint(expr= m.x5**2 + m.x13**2 + m.x21**2 == 1)
m.e6 = Constraint(expr= m.x6**2 + m.x14**2 + m.x22**2 == 1)
m.e7 = Constraint(expr= m.x7**2 + m.x15**2 + m.x23**2 == 1)
m.e8 = Constraint(expr= m.x8**2 + m.x16**2 + m.x24**2 == 1)
m.e9 = Constraint(expr= m.x17 - m.x18 >= 0)
m.e10 = Constraint(expr= m.x18 - m.x19 >= 0)
m.e11 = Constraint(expr= m.x19 - m.x20 >= 0)
m.e12 = Constraint(expr= m.x20 - m.x21 >= 0)
m.e13 = Constraint(expr= m.x21 - m.x22 >= 0)
m.e14 = Constraint(expr= m.x22 - m.x23 >= 0)
m.e15 = Constraint(expr= m.x23 - m.x24 >= 0)